Site search filtered paginated and info about no matches

peterbe opened the pull request:
Fixes #5658
Fixes #5748
Fixes #5808

joedarc commented on kuma/javascript/src/search-results-page.jsx:
```diff
@@ -1,24 +1,37 @@
 
-export type SearchResults = {
-    results: ?Array<{
+type SearchResultsResults = {
```
Not crazy about this name, maybe `SearchResults` here?

joedarc commented on kuma/javascript/src/search-results-page.jsx:
```diff
@@ -1,24 +1,37 @@
+    previous: ?string,
+    query: string
+};
+export type SearchResults = {
```
This could probably be `SearchResultsResponse` instead?

peterbe commented on kuma/javascript/src/search-results-page.jsx:
```diff
@@ -28,57 +41,153 @@ type Props = {
+            resultsNode = (results.documents || []).map(result => {
+                let path = `/${locale}/docs/${result.slug}`;
+                let url =
+                    window && window.origin ? `${window.origin}${path}` : path;
```
Because you don't have `window` in NodeJS in the SSR.

Gregoor replied:
It looks like there's already `BASE_URL` defined in this file for that purpose

peterbe replied:
There is no `BASE_URL` but the more I look at it I wonder why it isn't always just `'http://ssr.hack'`. The reason the `BASE_URL` exists is to be able to use `URL` and its tools for extracting searchParams and extracting a path etc.

peterbe commented on kuma/javascript/src/search-results-page.jsx:
```diff
@@ -109,18 +218,28 @@ export class SearchRoute extends Route<SearchRouteParams, SearchResults> {
         let path = parsed.pathname;
-        let q = parsed.searchParams.get('q');
+        let query = parsed.searchParams.get('q');
+        let page = parseInt(parsed.searchParams.get('page') || '1');
```
I asked [David](https://www.amazon.com/dp/0596805527), who wrote must of this code in the first place, about that and like the big dummy I am I can't remember why, but he convinced me it's not necessary. None of our `eslint` tooling is bothered by it.

joedarc replied:
Fair enough, just looking at [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#Octal_interpretations_with_no_radix) ES5 says that parseInt no longer allows ’0...’ Strings to be treated as octal values so that's probably safe.

peterbe replied:
We're not using ES5. We're using ES6, right? Which transpiles to ES5. Perhaps when it does so, it injects the optional "10" radix thing.

joedarc replied:
Right as long as the browser supports ES5 proper then the behavior for parseint has changed and this would be fine.

Gregoor replied:
I don't think babel (or at least the plugins we're using) does any non-syntactical transformations, so it shouldn't touch `parseInt`. But I also don't think we need-need the radix here, as the `page` shouldn't start with `0` or `0x`, so a radix of 10 should be assumed indeed.

peterbe commented:
Thank you @joedarc for your feedback. I really appreciate it.

peterbe commented:
@Gregoor I rebased this so it no longer conflicts on the git submodules. (no idea how that happened!)

Gregoor commented on kuma/api/v1/tests/test_views.py:
```diff
@@ -288,3 +289,48 @@ def test_whoami(user_client, api_settings, wiki_user, beta_testers_group,
+    assert response.status_code == 400
+    assert response.json()['error'] == "Search term 'q' must be set"
+
+    # 'q' present but locale invalid
```
This is imo a separate test case

peterbe replied:
I'm just lumping 3 checks into 1 python test function. I don't think is particularly important to break these apart into strict units because they're just not important enough.

Gregoor commented on kuma/javascript/src/search-results-page.jsx:
```diff
@@ -1,84 +1,206 @@
-                                    ))}
-                                </div>
+    // Because it's clunky to do conditionals in JSX, let's create all the
+    // React nodes here first.
```
I think it could be even less clunky, in terms of nesting and mismatch between declaration and insertion order. I'd personally go for more componentization.

peterbe replied:
Not sure what you mean. 
*Instead* doing:
```jsx
    if (!query) {
        noQuery = (
            <div className="result-container">
                <p>
                    <i>{gettext('Nothing found if nothing searched.')}</i>
                </p>
            </div>
        );
    }
```
it could be done like this:
```jsx
    if (!query) {
        noQuery = <NoQuery/>
    }
```
But I'm not sure it's of any use. It would make the total length-of-lines-of-code smaller for this render function but that means you'd have to jump between here go and find that `NoQuery` component. Now, although big, it reads naturally and you don't have to jump around in the file.

I personally like making components out of things to break them out, but only when it really helps. Such as, if it's used in more than 1 place. 

There was a point when this render function had a bunch of things like this:

```jsx
{data && !query && <STUFF>}
{data && results && !results.error && (results.previous || results.next) && <STUFF>}
{data && results && !results.error && !results.length && <STUFF>}
...etc...
```

It just was getting uglier and uglier. The lack of expressing simple `else` statements was annoying. 

I'm not in love with the way it turned out but it's not too far from good. At least it *reads* OK.

Gregoor replied:
I do agree with your view around componentization and not having to jump around in files generally. But I think in this case one has to jump anyway, since declaration and insertion are at different points. And imo jumping around for understanding content, is better than jumping around for understanding control flow. So my idea of more readable would look more like this:
```jsx
export default function SearchResultsPage({ locale, query, data }: Props) {
  const results = data ? data.results : null;
  return (
    <>
      <Header searchQuery={query} />
      {query && <Titlebar title={`${gettext('Results')}: ${query}`} />}

      <div className="search-results">
        {data && (
          <>
            {results && (
              <>
                <ResultsMeta {...{ results, locale }} />

                <Results {...{ results, locale }} />

                {(results.previous || results.next) && (
                  <Pager results={results} />
                )}

                {results.count === 0 && <NoResults />}
              </>
            )}
          </>
        )}

        {data.error && <ResultError error={data.error} />}

        {!query && (
          <div className="result-container">
            <p>
              <i>{gettext('Nothing found if nothing searched.')}</i>
            </p>
          </div>
        )}
      </div>
    </>
  );
}
```

Anyway, this might be more subjective than I think and I totally agree about merging and moving along :)

peterbe commented:
Note-to-self; I believe I've addressed the nits enough and I have the approval. If Travis goes green I'd rather merge this and move on than to debate nits. Not because I believe I'm right but because I believe the functionality isn't important enough to iron out nits. Does that make sense?

peterbe commented:
Actually, that looks pretty good! I didn't think about using those <>...</>
to make else statements.
If you take the time to copy this into a quick PR, I'd happily review it.

On Tue, Sep 24, 2019 at 5:57 AM Gregor <notifications@github.com> wrote:

> *@Gregoor* commented on this pull request.
> ------------------------------
>
> In kuma/javascript/src/search-results-page.jsx
> <https://github.com/mozilla/kuma/pull/5770#discussion_r327527715>:
>
> > -                                    </div>
> -                                    <div className="result-summary">
> -                                        {hit.summary}
> -                                    </div>
> -                                    {hit.excerpts.map((excerpt, i) => (
> -                                        <div
> -                                            className="result-excerpt"
> -                                            key={i}
> -                                            dangerouslySetInnerHTML={{
> -                                                __html: excerpt
> -                                            }}
> -                                        />
> -                                    ))}
> -                                </div>
> +    // Because it's clunky to do conditionals in JSX, let's create all the
> +    // React nodes here first.
>
> I do agree with your view around componentization and not having to jump
> around in files generally. But I think in this case one has to jump anyway,
> since declaration and insertion are at different points. And imo jumping
> around for understanding content, is better than jumping around for
> understanding control flow. So my idea of more readable would look more
> like this:
>
> export default function SearchResultsPage({ locale, query, data }: Props) {
>   const results = data ? data.results : null;
>   return (
>     <>
>       <Header searchQuery={query} />
>       {query && <Titlebar title={`${gettext('Results')}: ${query}`} />}
>
>       <div className="search-results">
>         {data && (
>           <>
>             {results && (
>               <>
>                 <ResultsMeta {...{ results, locale }} />
>
>                 <Results {...{ results, locale }} />
>
>                 {(results.previous || results.next) && (
>                   <Pager results={results} />
>                 )}
>
>                 {results.count === 0 && <NoResults />}
>               </>            )}          </>
>         )}
>
>         {data.error && <ResultError error={data.error} />}
>
>         {!query && (
>           <div className="result-container">
>             <p>
>               <i>{gettext('Nothing found if nothing searched.')}</i>
>             </p>
>           </div>
>         )}
>       </div>    </>
>   );}
>
> Anyway, this might be more subjective than I think and I totally agree
> about merging and moving along :)
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/kuma/pull/5770?email_source=notifications&email_token=AAAGQ4YOD5L2OGXXPGSRHYTQLHP6XA5CNFSM4IU3PMH2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFWDWRI#discussion_r327527715>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGQ44N7S7R3TPTFW6SBO3QLHP6XANCNFSM4IU3PMHQ>
> .
>
-- 
Peter Bengtsson
Mozilla MDN Web Docs
https://www.peterbe.com
